walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,207 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.lang.Math;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.Map;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Edge case if request's duration is over a day -> returns no options
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList();
+    }
+    // Get all mandatory meeting attendees & optional attendees
+    Set<String> mandatoryAttendees = new HashSet<>(request.getAttendees()); 
+    Set<String> optionalAttendees = new HashSet<>(request.getOptionalAttendees());
+    
+    // Process which events are mandatory to get required #
+    ArrayList<TimeRange> mandatoryEvents = new ArrayList<>();
+    for (Event i: events) {
+        if (mandatoryAttendees.containsAll(i.getAttendees())) {
+            mandatoryEvents.add(i.getWhen());
+        }
+    }
+
+    if (!(mandatoryAttendees.isEmpty()) && optionalAttendees.isEmpty()) {
+        return getTimeSlotWithMandatoryAttendeesOnly(mandatoryEvents, request, TimeRange.START_OF_DAY, TimeRange.END_OF_DAY);
+    } 
+    else if ((mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty())) { 
+        return fullDayAsOpenSlot();
+    } else {
+        return getTimeSlotWithMostOptionalAttendees(events, request, mandatoryAttendees, optionalAttendees, mandatoryEvents.size());
+    }
+  }
+
+  public Collection<TimeRange> getTimeSlotWithMostOptionalAttendees(Collection<Event> events, MeetingRequest request, 
```

Comment:
This function is really long and I think that splitting some of the code into helper functions would be helpful